### ISSUE TITLE
Performance of FEEvaluation::read_dof_values with masks

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#14098]
In https://github.com/dealii/dealii/pull/14085#issuecomment-1173676299 I noted that the matrix-vector product sees a slowdown of about 2% (consistent in terms of the statistical distribution), see also this image:
![step-37](https://user-images.githubusercontent.com/8276200/177502913-f059d27b-28ee-421e-b68d-b1e09d303e02.png)
Look at the dark blue and red lines around commit with index 870 or so. Furthermore, #14090 also adds a similar code. We should investigate the reason for the slow-down and fix it. It feels strange that such a small change can have an influence on the overall performance, but let's make sure we pick the best implementation in the end.

**Comment by kronbichler** [comment]
The fix in #14119 does improve things a bit, performance is now halfway between the regression of #14085 and the previous state.
![step-37](https://user-images.githubusercontent.com/8276200/179533132-134b0aea-bf6a-4dda-86e9-4b0be17b6291.png)

I think we should use this issue to collect some additional improvements in reading the values for continuous elements:
- The most obvious improvements is to use vectorized reads with masks for the zero-constrained case. This would mean adding a second `gather` and `scatter` function to signal a mask from the `indices` field, where we interpret `invalid_unsigned_int` as a masked integer.
- We should also use compressed index storage as we have done in a special project already: https://github.com/kronbichler/ceed_benchmarks_dealii/blob/experimental_merge_vector/common_code/vector_access_reduced.h

**Comment by kronbichler** [comment]
As a note to future work: Apart from the changes in #14314, there is a number of similar things to be fixed up in the `else` branch of the reads where we use scalar loads.

**Comment by kronbichler** [comment]
I think this will be addressed by #14324, so I added that PR on the milestone as well.

**Comment by kronbichler** [comment]
As discussed in https://github.com/dealii/dealii/pull/15673#issuecomment-1625151608, we should also check the operation count the case of DG elements with low polynomial degree and 2D.

**Comment by kronbichler** [comment]
This will be addressed by #14324 to such a good extent that I think we should consider closing this once that PR is merged. I will append this fact in the other PR to make sure this gets closed once done.
